### PR TITLE
Handle stale SSE connection on room join

### PR DESCRIPTION
**Pull request by kjamrog**
## Summary
- Add a library-level guard in SSE room join flow to skip stale/dead sessions before they are added to a room.
- Introduce a private close-and-unregister utility used by join guard and SSE error handling for consistent cleanup.
- Add/adjust tests for route utils, dual-mode keepAlive flow, and SSE logger typing updates.

## What changed
- `createRoomOperations(...)` now checks session liveness before `join`, logs a warning, and closes/unregisters dead sessions.
- `handleSSEError(...)` uses shared close/unregister behavior.
- Added e2e coverage to verify dead keepAlive sessions are not added to room membership and do not affect broadcasts.

## Summary by CodeRabbit

* **New Features**
  * Added warning-level logging support to SSE and DualMode loggers.

* **Bug Fixes**
  * Improved handling of dead or disconnected SSE sessions during room operations.
  * Enhanced error reporting for room subscription and unsubscription failures.

* **Tests**
  * Added comprehensive test coverage for edge cases involving disconnected sessions and room join scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/kibertoad/opinionated-machine/pull/162?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: lib/routes/fastifyRouteUtils.ts [hunk 1]
```diff
@@ -181,13 +181,7 @@ export async function handleSSEError(
...
   }
 
-  // Close the connection gracefully
-  try {
-    sseReply.sse.close()
-  } catch {
-    // Connection may already be closed
-  }
-  controller.unregisterConnection(connectionId)
+  closeSSESession(sseReply, controller, connectionId)
```
Comment (kibertoad) [inline lib/routes/fastifyRouteUtils.ts]: `closeSSESession` is now called here without a `logger` argument, but the new helper logs a warn via `logger?.warn(...)` when `sse.close()` throws. Net effect: error path stays silent, guard path becomes noisy — asymmetric. Either thread `options?.logger`/a logger argument through `handleSSEError` and pass it here, or drop the warn from `closeSSESession` and keep both call sites silent on double-close. Inconsistency will surprise future debuggers.

File: lib/routes/fastifyRouteUtils.ts [hunk 2]
```diff
@@ -216,10 +212,61 @@ function createRoomOperations(
...
+      if (isSSEConnectionDead(connection)) {
+        logger?.warn(
+          {
+            connectionId,
+            room,
+          },
+          'Skipping room join for dead SSE session',
+        )
+
+        closeSSESession(sseReply, controller, connectionId, logger)
```
Comment (kibertoad) [inline lib/routes/fastifyRouteUtils.ts:220]: When the guard fires, `closeSSESession` unregisters from the controller — but it doesn't fire the user-supplied `onClose(connection, 'server')` hook. The hook is registered via `sseReply.sse.onClose(...)` inside `start()`, which depends on the underlying stream emitting `close`. If `reply.raw` is already destroyed (the scenario this guard is built for), `sse.close()` will throw and the close-callback won't run. Application state set up in `onConnect` (e.g. event subscriptions, per-connection caches) will leak on dead-on-arrival sessions. Consider calling `callOnClose('server')` from this path, or at least documenting that `onConnect`/`onClose` are not symmetric in this corner case.

File: lib/routes/fastifyRouteUtils.ts [hunk 3]
```diff
@@ -216,10 +212,61 @@ function createRoomOperations(
...
+        return
+      }
+
+      roomManager.join(connectionId, room, logger)
+    },
+    leave: (room) => roomManager.leave(connectionId, room, logger),
+  }
+}
+
+export function isSSEConnectionDead<Events extends SSEEventSchemas, Context = unknown>(
```
Comment (kibertoad) [inline lib/routes/fastifyRouteUtils.ts]: Exporting `isSSEConnectionDead` (also re-exported from `lib/routes/index.ts`) makes the dead-state predicate part of the public API. Once external code depends on it, expanding or relaxing the set of checks becomes a versioning concern (e.g. adding a `socket.destroyed` check later could quietly change behavior). If consumers don't actually need this externally, prefer keeping it module-private and exposing a higher-level helper if needed.

File: lib/routes/fastifyRouteUtils.ts [hunk 4]
```diff
@@ -216,10 +212,61 @@ function createRoomOperations(
...
+function closeSSESession(
+  sseReply: SSEReply,
+  controller: SSEControllerLike,
+  connectionId: string,
+  logger?: SSELogger,
+): void {
+  try {
+    sseReply.sse.close()
+  } catch {
+    logger?.warn({ connectionId }, 'Connection is already closed')
```
Comment (kibertoad) [inline lib/routes/fastifyRouteUtils.ts]: Two issues:

1. The message asserts a cause ("Connection is already closed") that can't be known from a bare `catch` — `sse.close()` could throw for other reasons (programmer error, plugin bug). Phrase as "Error closing SSE stream" or similar.
2. The caught error is dropped. Include it: `logger?.warn({ err, connectionId }, 'Error closing SSE stream')`. Without `err`, this log is useless for debugging the non-trivial cases.

Also: with the current contract this fires every time we close an already-dead session via the guard path — that's expected, but means the warn channel will see steady traffic on a service with frequent client aborts. Consider whether `debug` level is more appropriate, or whether the warn is only useful at all for the unexpected-throw case.

File: lib/routes/fastifyRouteUtils.ts [hunk 5]
```diff
@@ -288,9 +333,16 @@ function createSSESessionInternal<Events extends SSEEventSchemas, Context = unkn
     isConnected: () => sseReply.sse.isConnected,
     getStream: () => sseReply.sse.stream(),
     sendStream,
-    rooms,
+    rooms: {
+      join: (_room) => {},
```
Comment (kibertoad) [inline lib/routes/fastifyRouteUtils.ts:336]: Minor code smell: `rooms` is built as a no-op placeholder here, then reassigned on line 343 by `createRoomOperations(connection, ...)`. The placeholder isn't observable (no async boundary between the two assignments) but the `_room` rename is a tell that this reads as dead code to a future reader / linter. A one-line comment explaining the circular-reference workaround (`createRoomOperations` closes over `connection`, so the object must exist before rooms can be wired up) would prevent someone from "cleaning it up" later.

File: lib/sse/rooms/SSERoomManager.ts [hunk 6]
```diff
@@ -105,8 +105,9 @@ export class SSERoomManager {
    *
    * @param connectionId - The connection to add to rooms
    * @param room - Room name or array of room names
+   * @param logger
```
Comment (kibertoad) [inline lib/sse/rooms/SSERoomManager.ts]: Empty `@param logger` — also at line 155 in `leave()`. Two words would do, e.g. "used to report adapter subscription failures." Without it, the new parameter looks accidental in the doc.

File: lib/sse/rooms/SSERoomManager.ts [hunk 7]
```diff
@@ -140,6 +141,7 @@ export class SSERoomManager {
       if (wasEmpty) {
         this.adapter.subscribe(r).catch(() => {
           // Log error but don't throw - subscription failure shouldn't break join
+          logger?.error({ connectionId, room: r }, 'Adapter subscription failed')
```
Comment (kibertoad) [inline lib/sse/rooms/SSERoomManager.ts]: The rejection reason from `adapter.subscribe(r)` is discarded — only `connectionId` and `room` make it into the log. When this fires in prod (e.g. Redis down), you want the actual error. Convert to `.catch((err) => { logger?.error({ err, connectionId, room: r }, 'Adapter subscription failed') })`. Same in the symmetric `leave` branch at line 179.

File: lib/routes/fastifyRouteUtils.spec.ts [hunk 8]
```diff
@@ -37,6 +42,141 @@ describe('fastifyRouteUtils', () => {
...
+        isConnected: () => true,
+        getStream: vi.fn(),
+        sendStream: vi.fn(),
+        rooms: { join: vi.fn(), leave: vi.fn() },
+      } as unknown as Parameters<typeof isSSEConnectionDead>[0]
+
+      expect(isSSEConnectionDead(session)).toBe(false)
+    })
+
+    it('marks session as dead when connection is closed', () => {
```
Comment (kibertoad) [inline lib/routes/fastifyRouteUtils.spec.ts]: `isSSEConnectionDead` has five OR conditions but only one (`isConnected() === false`) is exercised here. The other four — `request.raw.destroyed`, `request.raw.aborted`, `reply.raw.destroyed`, `reply.raw.writableEnded` — never get a positive case. A table-driven test (`it.each([...])`) covering each branch in isolation would catch a future regression where one of the conditions is accidentally dropped or inverted.

File: lib/routes/index.ts [hunk 9]
```diff
@@ -48,6 +48,7 @@ export {
   handleReconnection,
   handleSSEError,
   isErrorLike,
+  isSSEConnectionDead,
```
Comment (kibertoad) [inline lib/routes/index.ts]: Confirm exporting `isSSEConnectionDead` as part of the public API is intentional. Once consumers depend on it, the set of checks inside (currently 5 OR'd conditions) becomes a semver-relevant contract — expanding it later (e.g. adding `socket.destroyed`) could break their assumptions about what "dead" means. If no external consumer needs it today, keep it internal and add a higher-level helper if/when a use case arises.

**Review by kibertoad — COMMENTED** [review]
Reviewed the PR. Overall direction is solid — the guard correctly closes the window where a freshly-aborted connection can leave a phantom membership in a room. Below is a summary; specifics are inline.

**Endorsing CodeRabbit's findings:**

- `DualModeLogger.warn` and `SSELogger.warn` should be optional (`warn?:`). These types are deliberately minimal contracts for external consumers — making `warn` required is a source-level breaking change for anyone who hand-rolled a logger from just `error`. Every callsite already uses optional chaining (`logger?.warn(...)`), so the runtime side is already prepared for it. The `import 'pino'` shape isn't authoritative here; CommonLogger from `@lokalise/node-core` and any minimal mock have to compile too.
- `SSERoomManager.leave()` line 179 logs `"Adapter subscription failed"` from the `unsubscribe()` rejection — wrong verb. Same problem applies to the symmetric branch in `join()` line 144, except that one's actually correct; only `leave` has the misnomer.
- `simulateDeadBeforeJoin: z.coerce.boolean().optional()` in `testContracts.ts` is a footgun (string `"false"` coerces to `true`). Today the test only ever sets it to `true`, so it doesn't bite — but a future maintainer will hit it. Use `z.enum(['true','false']).transform(v => v === 'true')` or `z.stringbool()` if you're on Zod 4.

**Additional findings (inline):**

1. `handleSSEError` no longer passes a logger to `closeSSESession`, so the new "Connection is already closed" warn never fires from the error path — silent on one side, noisy on the other. Pick one behavior.
2. `closeSSESession`'s "Connection is already closed" log is presumptive — `sse.close()` can throw for other reasons, and the caught `err` is dropped on the floor. Capture it.
3. Same logger swallow problem in `SSERoomManager.join`/`leave` adapter failure branches: the rejection reason is discarded, only context is logged. If subscribe/unsubscribe genuinely fail, you want the stack.
4. Guard fires `closeSSESession` but the user's `onClose(connection, 'server')` lifecycle hook may not run for the dead-on-arrival case — `sse.close()` typically throws when raw is destroyed, so the `sseReply.sse.onClose(...)` callback registered in `start()` won't fire either. Application-level cleanup (subscriptions, per-connection state set in `onConnect`) will leak in this corner case. Worth at least documenting; ideally call `callOnClose('server')` from the guard path.
5. `isSSEConnectionDead` is now exported from `lib/routes/index.ts` — public API. Confirm that's intentional, since the dead-state semantic (which checks are included) becomes a versioning concern once external code depends on it.
6. `isSSEConnectionDead` unit coverage tests only the `isConnected() === false` branch. The other four conditions (`request.raw.destroyed`, `request.raw.aborted`, `reply.raw.destroyed`, `reply.raw.writableEnded`) are reachable but never exercised — easy table-driven test.
7. The `@param logger` JSDoc on `SSERoomManager.join`/`leave` is empty. Two-word description ("used to report adapter subscription failures") would be enough.
8. Minor code smell: `createSSESessionInternal` constructs `connection` with no-op `rooms`, then reassigns `connection.rooms`. The placeholder is never observed (no async between assignment and reassignment), but the `_room` parameter rename is a hint that this pattern reads as dead code. A short comment explaining the circular-reference workaround (`createRoomOperations` closes over `connection`) would help future readers.
9. The fix lives in `createRoomOperations` (route-utils layer), not at `SSERoomManager.join()`. That means any future caller bypassing route-utils and going directly to `roomManager.join(deadConnectionId, ...)` still re-introduces phantom memberships. Consider whether the dead check belongs lower in the stack — at minimum, document that `SSERoomManager.join` assumes the caller validated liveness.

**Not a blocker — possible follow-ups:**

- The TOCTOU window between `isSSEConnectionDead(connection)` and `roomManager.join(...)` is still nonzero; if the connection drops on the same tick, you'll add the membership and rely on `leaveAll` from the eventual close event to clean up. That's the existing behavior, so this PR doesn't regress it — but the commit message could note the limitation.
- UPGRADING.md / changelog: the dead-detection behavior is observable (sessions that previously joined rooms now log a warn and close); worth mentioning even though it's a bug fix.

File: lib/routes/fastifyRouteUtils.ts (hunk 2 above)
Comment (kjamrog) [reply lib/routes/fastifyRouteUtils.ts:220]: I'm not sure if `onClose` hook should be fired if connection is already closed. The most important part of the fix is skipping `roomManager.join()` for dead connections, rather than closing them automatically. 
It's true that calling `sse.close()` in most cases will have no effect now, but it's there mostly for consistency and being extra-safe, so I would say it's fine to skip the hook. Also, `callOnClose` is already preventing the hook from being called twice so in practice it won't have any effect anyway. 
What do you think?

Comment (kibertoad) [reply lib/routes/fastifyRouteUtils.ts:220]: yeah, you are right

File: lib/dualmode/dualModeTypes.ts [hunk 10]
```diff
@@ -9,6 +9,7 @@ export type DualModeType = 'json' | 'sse'
  */
 export type DualModeLogger = {
   error: (obj: Record<string, unknown>, msg: string) => void
+  warn: (obj: Record<string, unknown>, msg: string) => void
```
[the remark this answers is not in the document]
Comment (kjamrog) [reply lib/dualmode/dualModeTypes.ts]: Fixed, warn is optional now

File: lib/sse/sseTypes.ts [hunk 11]
```diff
@@ -24,6 +24,7 @@ export type AllContractEvents<Contracts extends Record<string, AnySSEContractDef
  */
 export type SSELogger = {
   error: (obj: Record<string, unknown>, msg: string) => void
+  warn: (obj: Record<string, unknown>, msg: string) => void
```
[the remark this answers is not in the document]
Comment (kjamrog) [reply lib/sse/sseTypes.ts]: Fixed, warn is optional now

File: test/dualmode/fixtures/testContracts.ts [hunk 12]
```diff
@@ -191,7 +191,10 @@ export const keepaliveDashboardContract = buildContract({
   method: 'get',
   pathResolver: () => '/api/dashboard/updates',
   requestPathParamsSchema: z.object({}),
-  requestQuerySchema: z.object({ dashboardId: z.string().optional() }),
+  requestQuerySchema: z.object({
+    dashboardId: z.string().optional(),
+    simulateDeadBeforeJoin: z.coerce.boolean().optional(),
```
[the remark this answers is not in the document]
Comment (kjamrog) [reply test/dualmode/fixtures/testContracts.ts]: Fixed, replaced with z.stringbool() and added test passing `simulateDeadBeforeJoin: 'false'` to confirm it's working as expected

File: lib/routes/fastifyRouteUtils.ts (hunk 1 above)
Comment (kjamrog) [reply lib/routes/fastifyRouteUtils.ts]: Added missing logger param

File: lib/routes/fastifyRouteUtils.ts (hunk 3 above)
Comment (kjamrog) [reply lib/routes/fastifyRouteUtils.ts]: Not needed to be exported for now - I made it private

File: lib/routes/index.ts (hunk 9 above)
Comment (kjamrog) [reply lib/routes/index.ts]: Not needed to be exported for now - I made it private

File: lib/routes/fastifyRouteUtils.ts (hunk 4 above)
Comment (kjamrog) [reply lib/routes/fastifyRouteUtils.ts]: Improved a bit - now we check if connection is still open in catch and only then we log `error`. Otherwise we skip the log

File: lib/sse/rooms/SSERoomManager.ts (hunk 7 above)
Comment (kjamrog) [reply lib/sse/rooms/SSERoomManager.ts]: Added error message to the log (not whole error to avoid issues with huge objects in logs)

File: lib/routes/fastifyRouteUtils.spec.ts (hunk 8 above)
Comment (kjamrog) [reply lib/routes/fastifyRouteUtils.spec.ts]: Done ✅

File: lib/sse/rooms/SSERoomManager.ts (hunk 6 above)
Comment (kjamrog) [reply lib/sse/rooms/SSERoomManager.ts]: Done ✅

**Review by kibertoad — APPROVED** [review]
